OpenRA.Mods.Common/Activities/Move/MoveAdjacentTo.cs

```diff
@@ -62,15 +61,33 @@ protected virtual bool ShouldRepath(Actor self, CPos targetLocation)
 			return lastVisibleTargetLocation != targetLocation;
 		}
 
-		protected virtual IEnumerable<CPos> CandidateMovementCells(Actor self)
+		protected override void OnFirstRun(Actor self)
 		{
-			return Util.AdjacentCells(self.World, Target)
-				.Where(c => Mobile.CanStayInCell(c));
+			QueueChild(Mobile.MoveTo(check => CalculatePathToTarget(self, check)));
 		}
 
-		protected override void OnFirstRun(Actor self)
+		protected readonly List<CPos> SearchCells = new List<CPos>();
+
+		protected int searchCellsTick = -1;
+		protected virtual bool UpdateSearchCells(Actor self)
```

Comment:
This doesn't look right, the method return value does not represent what the name suggests (whether the update succeeded or not). This suggests the method does more than it should: It should just update the cells and return void, while there should be another method (`CanSearchPath`?) that returns the boolean value.

Reply:
I agree it's a bit convoluted. This is quite a hot path and the problem is that, in `MoveAdjasentTo` and `MoveWithinRange` we only know wether we can path after updating search cells, and in MoveOnto, we know that before

Reply:
added a desctiption

Reply:
I think I will be able to make this function cleaner after #20767 is merged

Reply:
> I think I will be able to make this function cleaner after #20767 is merged

Adding the dependency label then.

Reply:
Cleaned up to how it would look like. It's slightly more duplicated logic but hopefully a much cleaner implementation and better readable